src/Maker/MakeAuthenticator.php

```diff
@@ -107,14 +166,62 @@ public function generate(InputInterface $input, ConsoleStyle $io, Generator $gen
                     $this->fileManager->getFileContents($path),
                     $input->getOption('firewall-name'),
                     $input->getOption('entry-point'),
-                    $classNameDetails->getFullName()
+                    $input->getArgument('authenticator-class')
                 );
                 $generator->dumpFile($path, $newYaml);
                 $securityYamlUpdated = true;
             } catch (YamlManipulationFailedException $e) {
             }
         }
 
+        if (self::AUTH_TYPE_FORM_LOGIN === $input->getArgument('authenticator-type')) {
+            $controllerClassNameDetails = $generator->createClassNameDetails(
+                $input->getArgument('controller-class'),
+                'Controller\\',
+                'Controller'
+            );
+
+            if (class_exists($controllerClassNameDetails->getFullName())) {
+                // If provided security controller class exist, add login() method
+                if (method_exists($controllerClassNameDetails->getFullName(), 'login')) {
+                    throw new RuntimeCommandException(sprintf('Method "login" already exists on class %s', $controllerClassNameDetails->getFullName()));
+                }
+
+                $manipulator = new ClassSourceManipulator(
+                    $this->fileManager->getFileContents($controllerPath = $this->fileManager->getRelativePathForFutureClass($controllerClassNameDetails->getFullName())),
+                    true
+                );
+                $securityControllerBuilder = new SecurityControllerBuilder();
+                $securityControllerBuilder->addLoginMethod($manipulator);
+
+                if (method_exists($controllerClassNameDetails->getFullName(), 'logout')) {
+                    throw new RuntimeCommandException(sprintf('Method "logout" already exists on class %s', $controllerClassNameDetails->getFullName()));
+                }
+                $securityControllerBuilder->addLogoutMethod($manipulator);
+
+                $this->generator->dumpFile($controllerPath, $manipulator->getSourceCode());
+            } else {
+                // otherwise, create security controller
+                $controllerPath = $generator->generateClass(
+                    $controllerClassNameDetails->getFullName(),
+                    'authenticator/SecurityController.tpl.php',
+                    [
+                        'parent_class_name' => \method_exists(AbstractController::class, 'getParameter') ? 'AbstractController' : 'Controller',
+                    ]
+                );
```

Comment:
We can simplify this a bit. You've already done the hard work in the method above to *modify* an existing controller... no matter what it looks like. So, here, we should just generate an *empty* class - something *almost* as simple as `Class.tpl.php` (but we'd obviously need the `extends` still). Then, we would run the code in the `if` part of this always. So:

```php
if (!class_exists($controllerClassNameDetails->getFullName())) {
    // make a dead-simple controller
}

// now, in all cases, do some ClassSourceManipulator magic!
```

Btw, I'm a bit concerned that we will now have the `parent_class_name` logic in 3 separate places. I'm not sure immediately how to fix that - but it bothers me.

Reply:
ok for generating an empty controller and then adding methods, indeed we got less duplicated code :)

about your concern, why don't we create a `Generator::generateController()` ?